### PR TITLE
[BE] feat: set interest topics for user

### PR DESCRIPTION
**Pull request by AngelC32**
- New feature implementation
- Established the basic file structure.
- Added entity userInterestsTopics
- Added UserInterestRepository interface for inserting new values into the database.
- Added interestTopicUseCase to use UserInterestRepository methods
- Added selectTopicHandler to handle HTTP requests
- Added unit tests

# Set interest topics for user
This feature allows the user to set their interest topics.

#### 1. **Endpoint**:

|Type|Route|
|-|-|
|POST|`https://{server}:{port}/interestTopics`|

#### 2. **Request Body**:

Field | Description | Example
------|-------------|--------
user_id | Current user ID| "1"
interest_id | Chosen topics ids array | ["1","2","3"]

#### 3. **Response Body**:

Field | Description | Example
----------|-------------|--------
response | Detailed task feedback | Insertion successful, Skipped setting interest topics, Error during insertion, Invalid request payload, Attempted insertion of an existing user interest topic
status | Status description | OK, ERROR

#### 4. Request example:

```json
{
    "user_id": "1",
    "interest_id": ["1","2","3"]
}
```

#### For successful cases:
```json
{
 "response": "Insertion successful",
 "status": "OK"
}
```
#### Skipped setting:
```json
{
 "response": "Skipped setting interest topics",
 "status": "OK"
}
```

#### Duplicate interest id:
```json
{
 "response": "Duplicate interest topic",
 "status": "ERROR"
}
```

#### Existing user interest topic:
```json
{
 "response": "Attempted insertion of an existing user interest topic",
 "status": "ERROR"
}
```
#### Invalid payload:
```json
{
 "response": "Invalid request payload",
 "status": "ERROR"
}
```

### CODE REVIEW

File: backend/config/config.go [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+
+type Config struct {
+	DBConnectionString string
+	AppPort            string
+}
+
+// Defines the connection string to the database
+func LoadConfig() *Config {
+	return &Config{
+		DBConnectionString: "host=localhost user=postgres password=root dbname=redsocial sslmode=disable",
```
Comment (eds000n) [inline backend/config/config.go:11]: we already have [this](https://github.com/unmsmfisi-socialapplication/social_app/blob/develop/backend/pkg/database/database.go), the PR must contain only code related to the feature

Comment (AngelC32) [reply backend/config/config.go]: You're right, the folder has been deleted.

**Comment by AngelC32** [comment]
## 📧 Set interest topics for user

### **API Overview**
Method: POST

### **API Components and Operation**

#### 1. **URL Schema**:
The endpoint to access the login service is structured as:
`https://{server}:{port}/interestTopics`

#### 2. **Request Body**:

Field | Description | Example
------|-------------|--------
user_id | Current user ID| "1”
interest_id | chosen topics ids array | [“1”,”2”,”3”]

#### 3. **Response Body**:

Field | Description | Example
----------|-------------|--------
response | Detailed task feedback | Insertion successful, Error during insertion, Invalid request payload, Attempted insertion of an existing user interest topic
status | Status description | OK, ERROR


Response examples:

#### For a successful cases:
```json
{
 "response": "Insertion successful",
 "status": "OK"
}
```

**Comment by AngelC32** [comment]
> @AngelC32 would you mind double checking the tests please? I changed de uri from `interestTopics` to `randomName`, basically the name could be change to whatever you want and the response will always be successful.
> 
> Is the web server being mocked?

**Comment by AngelC32** [comment]
> @AngelC32 would you mind double checking the tests please? I changed de uri from `interestTopics` to `randomName`, basically the name could be change to whatever you want and the response will always be successful.
> 
> Is the web server being mocked?

- You're right, I will be checking the tests.
- I was trying to mock the query to insert data into the respective table.

**Comment by AngelC32** [comment]
> > @AngelC32 would you mind double checking the tests please? I changed de uri from `interestTopics` to `randomName`, basically the name could be change to whatever you want and the response will always be successful.
> > Is the web server being mocked?
> 
> * You're right, I will be checking the tests.
> * I was trying to mock the query to insert data into the respective table.

@ekimike, the bug has been fixed.

File: backend/internal/interest_topics/application/interestTopicUseCase.go [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+	repo domain.UserInterestsRepository
+}
+
+func NewInterestTopicsUseCase(repo domain.UserInterestsRepository) *InterestTopicsUseCase {
+	return &InterestTopicsUseCase{
+		repo: repo,
+	}
+}
+
+func (itus *InterestTopicsUseCase) SetInterestTopics(user_id, interest_id string) error {
```
[the remark this answers is not in the document]
Comment (AngelC32) [reply backend/internal/interest_topics/application/interestTopicUseCase.go:27]: Well, I'm handling the insertion of multiple interest topics using a 'for' loop in the handler (selectTopicHandler).

File: backend/internal/interest_topics/infraestructure/selectTopicHandler.go [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+type SelectTopicHandler struct {
+	useCase application.InterestTopicsUseCaseI
+}
+
+func NewSelectTopicHandler(useCase application.InterestTopicsUseCaseI) *SelectTopicHandler {
+
+	return &SelectTopicHandler{useCase: useCase}
+}
+
+func (slh *SelectTopicHandler) HandleSelectTopic(w http.ResponseWriter, r *http.Request) {
```
[the remark this answers is not in the document]
Comment (AngelC32) [reply backend/internal/interest_topics/infraestructure/selectTopicHandler.go:21]: You're right, I'll rename those variables.

File: backend/internal/interest_topics/infraestructure/selectTopicHandler.go
[the remark this answers is not in the document]
Comment (AngelC32) [reply backend/internal/interest_topics/infraestructure/selectTopicHandler.go:1]: Ok, I'll make the changes

File: backend/internal/interest_topics/infraestructure/selectTopicHandler.go [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+				fmt.Println(err.Error())
+				return
+			default:
+				utils.SendJSONResponse(w, http.StatusNotFound, "NOTFOUND", "Error during insertion")
+				fmt.Println(err.Error())
+				return
+			}
+
+		}
+	}
```
[the remark this answers is not in the document]
Comment (AngelC32) [reply backend/internal/interest_topics/infraestructure/selectTopicHandler.go:89]: Thanks for your observations, I will be working on them. Actually, I hadn't been handling the atomicity of the transaction.

File: backend/internal/interest_topics/infraestructure/selectTopicHandler.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (AngelC32) [reply backend/internal/interest_topics/infraestructure/selectTopicHandler.go]: The variables were renamed in the following commit: [8d9944c](https://github.com/unmsmfisi-socialapplication/social_app/pull/151/commits/8d9944c5d58aa3a1a03ce9412b06fdbae7c1a93f)

File: backend/internal/interest_topics/application/interestTopicUseCase.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (AngelC32) [reply backend/internal/interest_topics/application/interestTopicUseCase.go]: Refactored code in this UseCase

File: backend/internal/interest_topics/infraestructure/selectTopicHandler.go
[the remark this answers is not in the document]
Comment (AngelC32) [reply backend/internal/interest_topics/infraestructure/selectTopicHandler.go:1]: I moved the "Invalid request payload" validations to a private function in the following commit: [8d9944c](https://github.com/unmsmfisi-socialapplication/social_app/pull/151/commits/8d9944c5d58aa3a1a03ce9412b06fdbae7c1a93f)

File: backend/internal/interest_topics/infraestructure/selectTopicHandler.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (AngelC32) [reply backend/internal/interest_topics/infraestructure/selectTopicHandler.go]: I tried to reduce calls to database, and considered transaction atomicity handling. Please, check it: [8d9944c](https://github.com/unmsmfisi-socialapplication/social_app/pull/151/commits/8d9944c5d58aa3a1a03ce9412b06fdbae7c1a93f)

File: backend/internal/interest_topics/infraestructure/selectTopicHandler.go
[the remark this answers is not in the document]
Comment (AngelC32) [reply backend/internal/interest_topics/infraestructure/selectTopicHandler.go:1]: Also the extra validations were moved to a private function in the last commit: [8e2cde7](https://github.com/unmsmfisi-socialapplication/social_app/pull/151/commits/8e2cde7cd7dd5bdf1b70d6b5d0516b1a02c4e8c7)

**Review by eds000n — APPROVED** [review]

**Comment by Cristhel17** [comment]
according to the QA contract for the user's topics of interest, the added format complies with the specifications of an identifier for the user or actor and a list to add the topics registered by user

**Pull request merged**